storage/invoke.go

```diff
@@ -27,6 +27,31 @@ import (
 	"google.golang.org/grpc/status"
 )
 
+// run determines whether a retry is necessary based on the config and
+// idempotency information. It then calls the function with or without retries
+// as appropriate, using the configured settings.
+func run(ctx context.Context, call func() error, retry *retryConfig, isIdempotent bool) error {
+	if !isIdempotent {
+		return call()
+	}
+	bo := gax.Backoff{}
+	if retry != nil && retry.backoff != nil {
+		bo.Multiplier = retry.backoff.Multiplier
+		bo.Initial = retry.backoff.Initial
+		bo.Max = retry.backoff.Max
+	}
+	return internal.Retry(ctx, bo, func() (stop bool, err error) {
+		err = call()
+		if err == nil {
+			return true, nil
+		}
+		if shouldRetry(err) {
+			return false, err
+		}
+		return true, err
```

Comment:
It doesn't explicitly state that, but maybe it should (or handle it instead of calling the `shouldRetry` func). 

The `invoke` code in gax-go exits early if `err == nil` ([here](https://github.com/googleapis/gax-go/blob/20dc16cacd1ba38bf7867c72797808cb3055f0e8/v2/invoke.go#L73)) but there is a lot more logic in that function than this one.

Reply:
Sounds good, yeah I think it is a reasonable thing to handle in any case so I made the change.

The current `shouldRetry` does eventually return false for a nil error as well but it's clearer to have an explicit short circuit.